### DISCUSSION TITLE
Clarification on message attributes

### DISCUSSION THREAD

**Discussion by MichaelBMiner** [cantools/cantools#450] [Q&A] [answered]
Hello,

I am working on a CAN parser and I am a bit confused when it comes to GenMsgSendType and GenMsgCycleTime, GenMsgDelayTime. 

In my DBC I have the following section. 
```
BA_DEF_ BO_  "GenMsgCycleTime" INT 0 0;
BA_DEF_ BO_  "GenMsgDelayTime" INT 0 0;
BA_DEF_ BO_  "GenMsgSendType" ENUM  "Cyclic","SendType1","SendType2","SendType3","SendType4","SendType5","SendType6","SendType7","SendType8","SendType9";
```

I assume this means I am only expecting cyclic style messages. Each message will come in at some specified time interval. If the time interval exceeds the GenMsgCycleTime, then I have an error. Is this correct?

If we assume SendType1 is instead Triggered, then would my message timeout rely on the GenMsgDelayTime?

**Comment by juleq** [comment]
Hi. These attributes are usually used to do a residual bus simulation. 

Example: You tell your tool: Please simulate node „ECU 1“. It will look into the send type and if it is cyclic, it will generate a new message every GenMsgCycleTime. You can offset the sending by GenMsgDelayTime to prevent unusual frame sending activity at one point in time.

These are attributes, so you could repurpose to your liking. Checking for missing frames using the cycle time is one example.

**Comment by juleq** [comment]
Some links with details on format:

http://mcu.so/Microcontroller/Automotive/DBC_File_Format_Documentation.pdf

https://github.com/stefanhoelzl/CANpy/blob/master/docs/DBC_Specification.md

I think there might be one more recent version of the Vector document but I could not find it quickly. It might have some additions.

**Comment by MichaelBMiner** [comment]
This is some great information. I am expecting to receive data cyclically from another system. My current criteria is to measure a timeout at 3x the GenMsgCycleTime without seeing the message. 

Thank you!

**Comment by juleq** [comment, marked as answer]
Welcome :). Sometimes this attribute might not have been defined. You should fall back to a sensible default then. Sometimes such a default is also defined by the dbc (attribute definition default). In that case, you could fall back to that.

In my experience the usage of these attributes is not very consistent across tools (because Vector did specify their existence, but to my knowledge not their purpose). I always resort to checking what Vector tools will do.
